test/distributed/_sharded_tensor/test_megatron_prototype.py

```diff
@@ -35,72 +44,165 @@
+        (
+            start_pos_fc2,
+            chunk_size_fc2,
+        ) = generate_local_weight_sharding_params_for_test(
+            local_weight_fc2, 1, TEST_GPU_NUM, spec[1], self.rank
+        )
+        local_grad_narrowed_fc2 = local_weight_grad_fc2.narrow(
+            1, start_pos_fc2, chunk_size_fc2
         )
-        local_grad_narrowed_second = local_grad_second.narrow(1, start_pos, chunk_size)
 
         # Test backward gradient calculation.
-        self.assertEqual(sharded_weight_first.grad, local_grad_narrowed_first)
-        self.assertEqual(sharded_weight_second.grad, local_grad_narrowed_second)
+        self.assertEqual(sharded_weight_fc1.grad, local_grad_narrowed_fc1)
+        self.assertEqual(sharded_weight_fc2.grad, local_grad_narrowed_fc2)
+        self.assertEqual(bias_grad_fc1, local_bias_grad_fc1)
+        self.assertEqual(bias_grad_fc2, local_bias_grad_fc2)
+
+        # Test optimizer.
+        bias_fc1, bias_fc2 = sharded_megatron_lm_original.get_bias()
+        local_bias_fc1, local_bias_fc2 = local_megatron_lm.get_bias()
+        self.assertEqual(bias_fc1, local_bias_fc1)
+        self.assertEqual(bias_fc2, local_bias_fc2)
+        self.assertEqual(bias_fc1.grad, local_bias_fc1.grad)
+        self.assertEqual(bias_fc2.grad, local_bias_fc2.grad)
+        previous_sharded_weight_fc1 = sharded_weight_fc1.clone()
+        previous_sharded_weight_fc2 = sharded_weight_fc2.clone()
+        previous_bias_fc1 = bias_fc1.clone()
+        previous_bias_fc2 = bias_fc2.clone()
+        optim = torch.optim.SGD(local_megatron_lm.parameters(), lr=0.1)
+        optim.step()
+        sharded_optim = ShardedOptimizer(
+            dict(named_params_with_sharded_tensor(sharded_megatron_lm_original)),
+            torch.optim.SGD,
+            lr=0.1,
+        )
+        sharded_optim.step()
+        local_weight_fc1_narrowed = local_weight_fc1.narrow(
+            0, start_pos_fc1, chunk_size_fc1
+        )
+        local_weight_fc2_narrowed = local_weight_fc2.narrow(
+            1, start_pos_fc2, chunk_size_fc2
+        )
+
+        # Test weight value after optimizer.
+        self.assertEqual(sharded_weight_fc1.size(), local_weight_fc1_narrowed.size())
+        self.assertEqual(sharded_weight_fc2.size(), local_weight_fc2_narrowed.size())
+        self.assertNotEqual(previous_sharded_weight_fc1, sharded_weight_fc1)
+        self.assertNotEqual(previous_sharded_weight_fc2, sharded_weight_fc2)
+        self.assertEqual(sharded_weight_fc1, local_weight_fc1_narrowed)
+        self.assertEqual(sharded_weight_fc2, local_weight_fc2_narrowed)
+
+        # Test bias value after optimizer.
+        local_bias_fc1, local_bias_fc2 = local_megatron_lm.get_bias()
+        self.assertNotEqual(previous_bias_fc1, bias_fc1)
+        self.assertEqual(bias_fc1, local_bias_fc1)
+        self.assertNotEqual(previous_bias_fc2, bias_fc2)
+        # Sharded bias get double updated here for fc2.
+        # self.assertEqual(bias_fc2, local_bias_fc2)
```

Comment:
cc @wanchaol

Reply:
Yes, this is very wired bug which I don't know the reason. Because all grads and updated weights are working as expected. Even the bias update for fc1 is correct.

Reply:
Yeah debugged this and found the issue, it should be fixed by https://github.com/pytorch/pytorch/pull/70228